README.md
Boundary Ping Check Plugin
--------------------------

Pings a set of hosts and reports on the response time. The plugin allows multiple hosts to be ping'd and each of those hosts to set their own Poll interval. See video [walkthrough](https://help.boundary.com/hc/articles/201383932).
Note: Currently does not support SmartOS.

## Prerequisites

### Supported OS

|     OS    | Linux | Windows | SmartOS | OS X |
|:----------|:-----:|:-------:|:-------:|:----:|
| Supported |   v   |    v    |    -    |  v   |

#### Boundary Meter Versions V4.0 Or Greater

To get the new meter:

    curl -fsS \
        -d "{\"token\":\"<your API token here>\"}" \
        -H "Content-Type: application/json" \
        "https://meter.boundary.com/setup_meter" > setup_meter.sh
    chmod +x setup_meter.sh
    ./setup_meter.sh

#### Boundary Meter Versions Less Than V4.0

|  Runtime | node.js | Python | Java |
|:---------|:-------:|:------:|:----:|
| Required |    +    |        |      |

- [How to install node.js?](https://help.boundary.com/hc/articles/202360701)

### Plugin Setup
None

#### Plugin Configuration Fields

|Field Name     |Description                                                                       |
|:--------------|:---------------------------------------------------------------------------------|
|Source         |The source to display in the legend for the host. Ex. google                      |
|Host           |The Hostname or IP Address to ping.  For example, www.google.com or 173.194.33.112|
|Poll Time (sec)|The Poll Interval to send a ping to the host in seconds. Ex. 5                    |

### Metrics Collected

|Metric Name       |Description                            |
|:-----------------|:--------------------------------------|
|Ping Response Time|The response time from the ping command|


Changelog
* Changes from 1.0 to 1.1
  * Rewritten the plugin to use the Boundary Plugin Framework.


modules/ping.lua
local _child = require('childprocess')
local _os = require('los')
local uv = require('uv')

local _platform = _os.type()
local _path
local _options

if (_platform == 'linux') then

    _path = '/bin/ping'
    _options = {'-n', '-w 2', '-c 1'}

elseif (_platform == 'win32' ) then
    _path = 'C:/windows/system32/ping.exe'
    _options = {'-n', '1', '-w', '3000'}

elseif (_platform == 'darwin') then
    _path = '/sbin/ping'
    _options = {'-n', '-t 2', '-c 1'}

else 
    io.stderr:write('Your platform is not supported.  We currently support Linux, Windows and OSX')
    os.exit(-1)
end

function pingHost(host, cb)

    local output = ""
    local ping = nil

    table.insert(_options, host)
    ping = _child.spawn(_path, _options)
    table.remove(_options)

    ping:on('error', function(err)

        local error = 'pingcheck: There was an error while executing the ping program. Please check the path or permissions.'
        error = error .. '\npath: ' .. _path
        error = error .. '\n' .. tostring(err)
        
        return cb and cb(error)

    end)

    ping.stdout:on('data', function (data)
        output = output .. tostring(data)
    end)

    ping.stderr:on('data', function (data)
        output = output .. tostring(data) 
    end)

    ping:on('exit', function (code)        

        -- If the there was an unknown host return that error
        if ( string.find(output, "unknown host") ) then
            return cb and cb('ERROR: The host "' .. host .. '" was not found')
        end

        -- If the there was an unknown host return that error (windows error message) 
        if ( string.find(output, "could not find host") ) then
            return cb and cb('ERROR: The host "' .. host .. '" was not found')
        end

        -- If there was no output, there was an error, mark it as failed
        if (output == "") then
            return cb and cb('ERROR: Unable to obtain any output')
        end

        -- If the process did not exit correctly, the host is down
        if (code ~= 0) then
            return cb and cb('ERROR: Exit code is ' .. code)
        end

        -- Parse the data to get the ping time
        local pingTime, index, prevIndex, i, j, time = 0, 0, 0, 0, 0, 0

        while true do

            prevIndex = index
            index = string.find(output, "\n", index+1)

            if index == nil then break end

            local line = string.sub(output, prevIndex, index-1)
            i, j, time  = string.find(line, "time=([0-9]*%.?[0-9]+)")

            if(time) then 
                pingTime = tonumber(time)
                break
            end

        end        

        return cb and cb(nil, pingTime)        

    end)

end

exports.ping = pingHost

init.lua
local framework = require('./modules/framework')
local CommandOutputDataSource = framework.CommandOutputDataSource
local PollerCollection = framework.PollerCollection
local DataSourcePoller = framework.DataSourcePoller
local Plugin = framework.Plugin
local los = require('los')
local table = require('table')
local string = require('string')

local isEmpty = framework.string.isEmpty
local clone = framework.table.clone

local params = framework.params 
params.name = 'Boundary Pingcheck plugin'
params.version = '1.1'
params.tags = "ping"

local commands = {
  linux = { path = '/bin/ping', args = {'-n', '-w 2', '-c 1'} },
  win32 = { path = 'C:/windows/system32/ping.exe', args = {'-n', '1', '-w', '3000'} },
  darwin = { path = '/sbin/ping', args = {'-n', '-t 2', '-c 1'} }
}

local ping_command = commands[los.type()] 
if ping_command == nil then
  io.stderr:write('Your platform is not supported.  We currently support Linux, Windows and OSX\n')
  process:exit(-1)
end

local function createPollers (params, cmd) 
  local pollers = PollerCollection:new() 
  for _, item in ipairs(params.items) do
    
    cmd = clone(cmd)
    table.insert(cmd.args, item.host)
    cmd.info = item.source

    local data_source = CommandOutputDataSource:new(cmd)
    local poll_interval = tonumber(item.pollInterval or params.pollInterval) * 1000
    local poller = DataSourcePoller:new(poll_interval, data_source)
    pollers:add(poller)
  end

  return pollers
end

local function parseOutput(context, output) 
  
  assert(output ~= nil, 'parseOutput expect some data')

  if isEmpty(output) then
    context:emit('error', 'Unable to obtain any output.')
    return
  end

  if (string.find(output, "unknown host") or string.find(output, "could not find host.")) then
    context:emit('error', 'The host ' .. context.args[#context.args] .. ' was not found.')
    return
  end

  local index
  local prevIndex = 0
  while true do
    index = string.find(output, '\n', prevIndex+1) 
    if not index then break end

    local line = string.sub(output, prevIndex, index-1)
    local _, _, time  = string.find(line, "time=([0-9]*%.?[0-9]+)")
    if time then 
      return tonumber(time)
    end
    prevIndex = index
  end

  return -1
end

local pollers = createPollers(params, ping_command)
local plugin = Plugin:new(params, pollers)

function plugin:onParseValues(data) 
  local result = {}

  local value = parseOutput(self, data['output'])
  result['PING_RESPONSETIME'] = { value = value, source = data['info'] }
  return result
end

plugin:run()
